### ISSUE TITLE
Sometimes uploads fails or queued, then don't upload.

### ISSUE THREAD

**Issue by neslihanturan** [commons-app/apps-android-commons#1878]
**Summary:** 

Our friend John Lubbock says, "I wonder if you know why when I upload photos on the Commons app, sometimes they say 'Failed' or 'queued', and don't upload?".

**Device and Android version:** 
Samsung S9+
 
 **Commons app version:** 

2.8.2

**Screen-shots:** 
![image](https://user-images.githubusercontent.com/3127881/45031645-17572d80-b058-11e8-89c6-85165f33dca6.png)

**Would you like to work on the issue?**
^_^ no.

**Comment by misaochan** [comment]
Where did John post, @neslihanturan ? I'd like to ask him about the method he used to upload (Nearby? Share? etc)

**Comment by neslihanturan** [comment]
I asked same question too, will update once the answer comes. He asked from twitter, I don't think he has a github account.

**Comment by neslihanturan** [comment]
He said contributions were from main screen.

**Comment by misaochan** [comment]
Could you point us to the conversation on Twitter please?

**Comment by misaochan** [comment]
Sorry, I'm not familiar with Twitter and didn't know you could send private messages on it - I thought he tweeted it, haha. If it's a private message then never mind. Could you ask him how often it happens for him, and if the queued pictures ever succeeded? Also, he has confirmed he is on 2.8.2?

**Comment by neslihanturan** [comment]
- Yes he confirmed he is on 2.8.2 with about page screenshoot. 
- He said uploads don't succeed afterwards.
- He said sometimes.

**Comment by misaochan** [comment]
Yann posted in beta feedback:

> Some files stay in the queue. I can't upload or remove them, except by uninstalling the application. File chosen from the gallery on Android. It happens repeatedly if a file already exists with the same name. AFAIK it only happens from 2.8. It happens even if I upload one file at a time.

Do you think the changes that we made to file uploads in 2.8 might be the culprit? @neslihanturan could you please ask John if he has a rough idea of when the problem started for him?

**Comment by misaochan** [comment]
Wanted to bump this issue as we have another report on google groups:

```
Picture just ends up as queued.

The only thing shown is the file name all the other items are blank even though I filled them in. Anyway the whole app doesn't work here on Android 8.0. Bye.
```

I guess the main question at the moment is whether this issue can conceivably be solved in a reasonable time before our upload backend overhaul in next year's PG proposal. It seems to be a high priority issue, but we are already behind on all of our deadlines.

**Comment by misaochan** [comment]
Yann has informed me that v2.8.4 does not solve this problem for him. I have been talking to [Jim Henderson](https://commons.wikimedia.org/wiki/Commons_talk:Mobile_app#Crash_on_upload), who reports a similar issue with queued uploads.

@neslihanturan , after you are done with your main UI PR, could you please go through your PR with the file upload changes, and see if anything that you changed could have caused this? 

I suggest you try with an Android 8.0 or 8.1 emulator. So far all of the reports appear to be from that version. I just need to confirm with Jim about his Android version

**Comment by Yannf** [comment]
FYI my Android version is 7.1.1, and it is the latest update on my J5. Thanks for trying to fix this, it is rather annoying.

**Comment by misaochan** [comment]
@Yannf Thanks for chiming in! Could you please elaborate on what you mean by "first picture in a series", by detailing the steps you take and the results? E.g.

1. Open app, tapped gallery button, select picture, enter title/desc etc. Picture ends up queued forever.
2. In app, tapped gallery button again shortly after that and selected another picture, entered title/desc. Picture upload succeeds

So on and so forth?

**Comment by Yannf** [comment]
If I upload several pictures successively, usually only the first fails. And it seems the App crashes (it closes without warning).

**Comment by ashishkumar468** [comment]
@misaochan Will take this up

**Comment by ashishkumar468** [comment]
I was actually trying to debug this issue on Google Pixel. Seems like something is wrong with the api's.
So the issue is that while trying to upload an image from google photos app, (ie. share photo). Although the MIME type received in the intent is jpeg for which we send jpg to our backend, we get an error response `<?xml version="1.0" encoding="UTF-8"?>
<api servedby="deployment-mediawiki-09">
   <error code="verification-error" info="File extension &quot;.jpg&quot; does not match the detected MIME type of the file (image/png).">
      <details>
         <detail>filetype-mime-mismatch</detail>
         <detail>jpg</detail>
         <detail>image/png</detail>
      </details>
      <docref xml:space="preserve">See https://commons.wikimedia.beta.wmflabs.org/w/api.php for API usage. Subscribe to the mediawiki-api-announce mailing list at &amp;lt;https://lists.wikimedia.org/mailman/listinfo/mediawiki-api-announce&amp;gt; for notice of API deprecations and breaking changes.</docref>
   </error>
</api>`
 
which states that it was expecting png while we sent the file as jpeg. I tried hard coding the image extension to png and the upload succeeded [Shocking]. I could add an exception where in I could check if the photos are uploaded from google photos make them png by default. Doesnot seems right but would fix the upload fail cases from google photos. Also, from the code i can conclude that this should be happening since quite a while as say the comments as well, TODOing to add fix for uploads via Google Photos,. @misaochan @neslihanturan  What do you guys suggest ?

**Comment by ashishkumar468** [comment]
Actually not a backend bug, Although in the image info, Google Photo shows that the image is png but while sharing it, in the intent it shows that the image is jpeg and hence our backend rejects that because of which the upload fails

**Comment by neslihanturan** [comment]
@ashishkumar468 thanks for digging it! I think you are very close since we recently implemented some changes about file extensions. We convert some extensions to .jpeg or .jpg (I am not really sure but can check).

**Comment by ashishkumar468** [comment]
@neslihanturan Yes, we convert jpeg to jpg  but that is not causing any trouble. The issue is that google photos app is giving the mime type for png images as jpeg which is causing the api to fail.

**Comment by Yannf** [comment]
Hi,

On Tue, 9 Oct 2018, 00:29 Ashish Kumar, <notifications@github.com> wrote:

> @neslihanturan <https://github.com/neslihanturan> Yes, we convert jpeg to
> jpg but that is not causing any trouble. The issue is that google photos
> app is giving the mime type for png images as jpeg which is causing the api
> to fail.
>
But it only happens sometimes, and it seems in specific circumstances.

Yann Forget

>

**Comment by ashishkumar468** [comment]
@Yannf Yes and that is what is causing the api's to fail at least in my case. AFAIK, upload does not fail always, right ?

**Comment by Yannf** [comment]
No, not always.
I checked my software. I am using the default Gallery application version
5.4.03.2 on Android, in case it matters.

Yann Forget

**Comment by ashishkumar468** [comment]
@Yannf  and the uploads are failing in that as well ?. It would be great if you could attach the logs ?

**Comment by ashishkumar468** [comment]
@misaochan  The main job would be to validate the file extension somehow locally. As I have mentioned earlier, even if the extensions are png google photos are giving the MIME type as jpeg. We will have to figure out a way to validate file extensions. Apart from this indefinite  progress in notification failed uploads, happening on my device due to the same reason and also faced by Jim as well would be resolved if we are able to add a patch for this.

**Comment by misaochan** [comment]
Yes, we certainly need to check the file extension locally. I would be surprised if the error lies in the Google Photos API though, as I haven't been able to find any reports of that. I would have thought that it is our app erroneously assigning the MIME type as .jpeg, due to the hack I mentioned above.

**Comment by ashishkumar468** [comment]
@misaochan  @neslihanturan  I have added a fix for MIME type detection, although could not find a logical bug in the method which actually attaches the extension. Although the flow needs to be refactored a bit which we can take later on. Seems like this will add a fix for the image extension issue. Raising a PR for the same

**Comment by ashishkumar468** [comment]
Had a basic question although , do the releases not go through master branch, as I could conclude from the thread that the issue went through the branch 2.8-release. If so, which branch am I supposed to make the PR to ? I had solved the issue on a branch cut from 2.8-release. Do I need to do it from master. It would be great if some one could clarify the same
